### PR TITLE
feat(Traits) add a number of test case traits

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://moderntribe.atlassian.net/browse/TEC-3225

This pr adds the following traits:
* `With_Ascii_Calendar` to render an ASCII calendar for easier, visual,
human compatible debugging.
* `With_Json_Data_Sets` to allow loading data sets, for data providers,
from JSON files, in place of specifying them in line in the test case.

Furhtermore this pr adds a number of support methods to the
`With_View_Context` trait.

Example ASCII calendar render:
ascii-calendar

### CODE REVIEW

**Comment by lucatume** [comment]
To have an example of usage: https://github.com/moderntribe/the-events-calendar/pull/3035

File: src/Products/Traits/With_View_Context.php [hunk 1]
```diff
@@ -58,6 +61,115 @@ protected function setup_context( array $alterations = [] ) {
...
+		return "Event legend:\n\n" . implode(
+			"\n",
+			array_map(
+				static function ( $event ) {
+					return sprintf(
+						'%d - tz: %s; start: %s; end: %s; all-day: %s',
+						$event->ID,
+						get_post_meta( $event->ID, '_EventTimezone', true ),
+						get_post_meta( $event->ID, '_EventStartDate', true ),
+						get_post_meta( $event->ID, '_EventEndDate', true ),
```
Comment (bordoni) [inline src/Products/Traits/With_View_Context.php]: Should we have something for UTC as well?

File: src/Products/Traits/With_Json_Data_Sets.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	 * @return \Generator The decoded JSON data sets.
+	 *
+	 * @throws \RuntimeException If the data set directory does not exist and cannot be created.
+	 */
+	public function json_data_sets( $test_method = null ) {
+		$r   = new \ReflectionClass( $this );
+		$dir = implode( '/', [ dirname( $r->getFileName() ), 'data-sets', $r->getShortName() ] );
+
+		if ( ! ( is_dir( $dir ) || mkdir( $dir, 0777, true ) ) ) {
+			throw new \RuntimeException( "Cannot create data-set dir: {$dir}" );
```
Comment (Camwyn) [inline src/Products/Traits/With_Json_Data_Sets.php]: Not sure why we're creating the folder here?
There's nothing in the docblock about that, and this function doesn't generate data, just reads/formats it, correct?

File: src/Products/Traits/With_View_Context.php (hunk 1 above)
Comment (lucatume) [reply src/Products/Traits/With_View_Context.php]: I did not want to burden the legend.
Unless we really need to, I would not add it.

File: src/Products/Traits/With_Json_Data_Sets.php (hunk 2 above)
Comment (lucatume) [reply src/Products/Traits/With_Json_Data_Sets.php]: This function only reads the data, correct.

Yet the expected location of files is not immediate from the code and I just wanted to lower the debug barrier ("Where should I store the JSON data sets to have them read here?!") by doing it.
It's pretty common to scaffold and read to speed up test development.

**Comment by lucatume** [comment]
I've refactored the ASCII calendar code into a stand-alone, and re-usable, class.
I've also broken down the code for readability's sake and expose more parameters on it.
Finally, I've improved the code to keep events aligned across days:
ascii-calendar-2 0

File: src/Products/Traits/With_Json_Data_Sets.php (hunk 2 above)
Comment (lucatume) [reply src/Products/Traits/With_Json_Data_Sets.php]: I'm adding a note about this in the code block.

**Pull request merged**